### PR TITLE
[TOPIC-BT-LLCP] Bluetooth: controller: llcp: Feature Exchange control procedure

### PR DESCRIPTION
**Pull request by kruithofa**
Implements the feature exchange control procedure and unit tests

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

### CODE REVIEW

File: tests/bluetooth/controller/common/src/helper_pdu.c [hunk 1]
```diff
@@ -33,8 +33,47 @@
...
 
+void helper_pdu_encode_feature_req(struct pdu_data *pdu, void *param)
+{
+	pdu->ll_id = PDU_DATA_LLID_CTRL;
+	pdu->len = offsetof(struct pdu_data_llctrl, feature_req) +
+		sizeof(struct pdu_data_llctrl_feature_req);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_FEATURE_REQ;
+	for (int counter = 0; counter < 8; counter++) {
+		u8_t expected_value =
+			((u64_t) EXPECTED_FEATURE >> 8*counter) & 0xFF;
```
Comment (thoh-ot) [inline tests/bluetooth/controller/common/src/helper_pdu.c:47]: Not liking this approach as we cannot control FeatureSet_S and FeatureSet_M to verify FeatureSet_USED logic. Same in all the other places where EXPECTED_FEATURE is used.

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+ *    |                            |                   |
+ *    |                            |    LL_FEATURE_RSP |
+ *    |                            |<------------------|
+ *    |                            |                   |
+ *    |     Feature Exchange Proc. |                   |
+ *    |                   Complete |                   |
+ *    |<---------------------------|                   |
+ *    |                            |                   |
+ */
+void test_feature_exchange_mas_loc(void)
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:75]: No control over FeatureSet_S and FeatureSet_M to verify FeatureSet_USED logic.

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+	/* There should be one host notification */
+
+	ut_rx_pdu(LL_FEATURE_RSP, &ntf,  &remote_feature_rsp);
+	ut_rx_q_is_empty();
+	zassert_equal(conn.lll.event_counter, 1,
+		      "Wrong event-count %d\n", conn.lll.event_counter);
+}
+
+
+void test_feature_exchange_mas_loc_unknown_rsp(void)
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:234]: Not making sense; the LL_UNKNOWN_RSP is not valid for LL_FEATURE_REQ only for LL_SLAVE_FEATURE_REQ.

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+	zassert_equal(conn.lll.event_counter, 0,
+		      "Wrong event-count %d\n", conn.lll.event_counter);
+}
+
+/*
+ * +-----+ +-------+            +-----+
+ * | UT  | | LL_A  |            | LT  |
+ * +-----+ +-------+            +-----+
+ *   |        |                   |
+ *   |        |    LL_FEATURE_REQ |
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:134]: This should say LL_SLAVE_FEATURE_REQ.

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+	event_done(&conn);
+	err = ull_cp_feature_exchange(&conn);
+	zassert_equal(err, BT_HCI_ERR_SUCCESS, NULL);
+
+	event_prepare(&conn);
+	lt_rx(LL_FEATURE_RSP, &conn, &tx, &local_feature_rsp);
+	lt_rx_q_is_empty(&conn);
+	event_done(&conn);
+
+	ut_rx_pdu(LL_SLAVE_FEATURE_REQ, &ntf, &remote_feature_req);
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:192]: Not a valid PDU towards the host.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 6]
```diff
@@ -127,10 +131,19 @@ static void lp_comm_ntf(struct ull_cp_conn *conn, struct proc_ctx *ctx)
...
 
-	switch (ctx->proc) {
-	case PROC_VERSION_EXCHANGE:
+	switch (conn->response_opcode) {
+	case PDU_DATA_LLCTRL_TYPE_FEATURE_RSP:
+		ntf_encode_feature_rsp(conn, pdu);
+		break;
+	case PDU_DATA_LLCTRL_TYPE_SLAVE_FEATURE_REQ:
+		ntf_encode_slave_feature_req(conn, pdu);
+		break;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:140]: Not a valid host notification.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 7]
```diff
@@ -127,10 +131,19 @@ static void lp_comm_ntf(struct ull_cp_conn *conn, struct proc_ctx *ctx)
 	ntf->hdr.type = NODE_RX_TYPE_DC_PDU;
 	pdu = (struct pdu_data *) ntf->pdu;
 
-	switch (ctx->proc) {
-	case PROC_VERSION_EXCHANGE:
+	switch (conn->response_opcode) {
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:134]: Not sure i'm in favor of this.. think it's dangerous to tie host notification up on the received pdu and not on the active procedure.

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h [hunk 8]
```diff
@@ -9,7 +9,7 @@
  */
 struct ull_cp_conn {
 	struct ull_tx_q tx_q;
-
+	enum pdu_data_llctrl_type response_opcode;
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.h:12]: good point

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 9]
```diff
@@ -273,10 +321,22 @@ void ull_cp_priv_lp_comm_run(struct ull_cp_conn *conn, struct proc_ctx *ctx, voi
...
+	case PDU_DATA_LLCTRL_TYPE_FEATURE_REQ:
+	case PDU_DATA_LLCTRL_TYPE_SLAVE_FEATURE_REQ:
+		pdu_decode_feature_req(conn, pdu);
+		break;
 	case PDU_DATA_LLCTRL_TYPE_VERSION_IND:
 		pdu_decode_version_ind(conn, pdu);
 		break;
+	case PDU_DATA_LLCTRL_TYPE_UNKNOWN_RSP:
+		pdu_decode_unknown_rsp(conn, pdu);
+		break;
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:339]: removed

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 10]
```diff
@@ -326,6 +390,17 @@ static void rp_comm_st_idle(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t
 static void rp_comm_send_rsp(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
 {
 	switch (ctx->proc) {
+	case PROC_FEATURE_EXCHANGE:
+		/* We allow multipe times a feature exchange */
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:394]: done

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 11]
```diff
@@ -273,10 +321,22 @@ void ull_cp_priv_lp_comm_run(struct ull_cp_conn *conn, struct proc_ctx *ctx, voi
 
 static void rp_comm_rx_decode(struct ull_cp_conn *conn, struct pdu_data *pdu)
 {
+	conn->response_opcode = pdu->llctrl.opcode;
+
 	switch (pdu->llctrl.opcode) {
+	case PDU_DATA_LLCTRL_TYPE_FEATURE_RSP:
+		pdu_decode_feature_rsp(conn, pdu);
+		break;
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:329]: ok

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 3 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:234]: changed accordingly

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 4 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:134]: done

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 5 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:192]: changed

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c (hunk 6 above)
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:140]: modified to normal feature_req

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c (hunk 7 above)
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:134]: Hm, good point, probably we should keep the check for proper procedure, but we'll still have to  check that we received the proper PDU type back before decoding it. I'll think of something

Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: Ok, added some better checking here

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+	struct node_tx *tx;
+
+
+	struct pdu_data_llctrl_feature_req local_feature_req;
+	/*
+	 * 	struct node_rx_pdu *ntf;
+	 *	struct pdu_data_llctrl_unknown_rsp unknown_rsp = {
+	 *		.type = PDU_DATA_LLCTRL_TYPE_FEATURE_REQ
+	 *	};
+	 */
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:257]: Why is this here?

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+	 * lt_tx(LL_UNKNOWN_RSP, &conn, &unknown_rsp);
+	 */
+	event_done(&conn);
+
+	/*
+	 * ut_rx_pdu(LL_UNKNOWN_RSP, &ntf,  &unknown_rsp);
+	 * ut_rx_q_is_empty();
+	 * zassert_equal(conn.lll.event_counter, 1,
+	 * 	      "Wrong event-count %d\n", conn.lll.event_counter);
+	 */
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:287]: Why is this here?

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+
+	/* Tx Queue should have one LL Control PDU */
+	lt_rx(LL_SLAVE_FEATURE_REQ, &conn, &tx, &local_feature_req);
+	lt_rx_q_is_empty(&conn);
+
+
+	/* Rx Commented out for know, handling of UNKNOWN response will come in an update */
+	/*
+	 * lt_tx(LL_UNKNOWN_RSP, &conn, &unknown_rsp);
+	 */
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:279]: Why is this here?

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: tests/bluetooth/controller/common/defaults_cmake.txt [hunk 15]
```diff
@@ -29,12 +29,15 @@ FILE(GLOB ll_sw_sources
...
+        $ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ll_assert.c
 )
 
 FILE(GLOB common_sources
         $ENV{ZEPHYR_BASE}/tests/bluetooth/controller/common/src/helper_pdu.c
         $ENV{ZEPHYR_BASE}/tests/bluetooth/controller/common/src/helper_util.c
 )
 
+set(CMAKE_BUILD_TYPE Debug)
+
```
Comment (thoh-ot) [inline tests/bluetooth/controller/common/defaults_cmake.txt:41]: You can run sanitycheck with -xCMAKE_BUILD_TYPE=Debug

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 16]
```diff
@@ -115,6 +119,43 @@ static void lp_comm_tx(struct ull_cp_conn *conn, struct proc_ctx *ctx)
...
+	case PDU_DATA_LLCTRL_TYPE_VERSION_IND:
+		ntf_encode_version_ind(conn, pdu);
+		break;
+	default:
+		/* EGON: TODO: define behaviour for unexpected PDU */
+		LL_ASSERT(0);
+	}
+}
+
+static void lp_comm_ntf_unknown_proc(struct ull_cp_conn *conn,  struct proc_ctx *ctx, struct pdu_data *pdu)
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: Removed the complete procedure since it is not used; I had hoped to implement Unknown Response, but that requires some more work, in particular for testing it

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h [hunk 17]
```diff
@@ -7,9 +7,11 @@
 /* LLCP Procedure */
 enum llcp_proc {
 	PROC_UNKNOWN,
+	PROC_FEATURE_EXCHANGE,
 	PROC_VERSION_EXCHANGE,
 	PROC_ENCRYPTION_START,
 	PROC_PHY_UPDATE,
+	PROC_UNKNOWN_RESP,
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h]: Removed, I wanted to implement sending Unknown Response, but that requires more work for testing

File: subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c [hunk 18]
```diff
@@ -28,12 +28,143 @@
...
+	struct pdu_data_llctrl_feature_rsp *p;
+
+	pdu->ll_id = PDU_DATA_LLID_CTRL;
+	pdu->len = offsetof(struct pdu_data_llctrl, feature_rsp) +
+		sizeof(struct pdu_data_llctrl_feature_rsp);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_FEATURE_RSP;
+	p = &pdu->llctrl.feature_rsp;
+
+	pdu_encode_set_features(conn, p->features);
+//	sys_put_le64(conn->llcp.fex.features, p->features);
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c]: Done

File: subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c [hunk 19]
```diff
@@ -409,8 +421,8 @@ void ull_cp_priv_rr_new(struct ull_cp_conn *conn, struct node_rx_pdu *rx)
 		proc = PROC_PHY_UPDATE;
 		break;
 	default:
-		/* Unknown opcode */
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:412]: Put back

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 12 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c]: For testing unknown response; I had some issues, but that's now fixed

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 13 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c]: see above

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 14 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c]: see above

File: tests/bluetooth/controller/common/defaults_cmake.txt (hunk 15 above)
Comment (kruithofa) [reply tests/bluetooth/controller/common/defaults_cmake.txt]: Ok

File: tests/bluetooth/controller/common/src/helper_pdu.c (hunk 1 above)
Comment (kruithofa) [reply tests/bluetooth/controller/common/src/helper_pdu.c:47]: I'll pass it in as a parameter, then we can control it from the unittests

File: tests/bluetooth/controller/ctrl_feature_exchange/src/main.c (hunk 2 above)
Comment (kruithofa) [reply tests/bluetooth/controller/ctrl_feature_exchange/src/main.c:75]: Features are now passed in as parameter

**Review by thoh-ot — APPROVED** [review]
